mocks/DDsmu_mocks/countpairs_s_mu_mocks_kernels.c.src

```diff
@@ -0,0 +1,884 @@
+    for(int64_t i=0;i<N0;i++) {
+        const DOUBLE xpos = *x0++;
+        const DOUBLE ypos = *y0++;
+        const DOUBLE zpos = *z0++;
+        const DOUBLE dpos = *d0++;//d is the co-moving distance
+        for(int w = 0; w < pair.num_weights; w++){
+            pair.weights0[w].d = *local_w0.weights[w]++;
+        }
+
+        /* If in the same cell, unique pairs are guaranteed by not including the current particle */
+        if(same_cell == 1) {
+            d1++; n_off++;
+            nleft--;
+        } else {
+            /* For a different cell, all pairs are unique pairs, since two cells are only opened for pairs once (accounted for in the assign_ngb_cells function)*/
+            while(nleft > 0) {
+                /*Particles are sorted on 'd', in increasing order */
+                const DOUBLE dz = *d1 - dpos;
+                if(dz > -smax) break;
+                d1++; n_off++;
+                nleft--;
+            }
+            /*If no particle in the second cell satisfies distance constraints on 'dz' for the current 'i'th particle in first cell,
+              then there can be no more pairs from any particles in the first cell (since the first cell is also sorted in increasing order in 'd')
+             */
+            if(nleft == 0) {
+                i=N0;
+                break;
+            }
+        }
+
+        DOUBLE *localx1 = x1 + n_off;
+        DOUBLE *localy1 = y1 + n_off;
+        DOUBLE *localz1 = z1 + n_off;
+        DOUBLE *locald1 = d1;
+        for(int w = 0; w < pair.num_weights; w++){
+            local_w1.weights[w] = weights1->weights[w] + n_off;
+        }
+
+        for(int64_t j=0;j<nleft;j++){
+            const DOUBLE parx = xpos + *localx1;
+            const DOUBLE pary = ypos + *localy1;
+            const DOUBLE parz = zpos + *localz1;
+
+            const DOUBLE perpx = xpos - *localx1;
+            const DOUBLE perpy = ypos - *localy1;
+            const DOUBLE perpz = zpos - *localz1;
+
+            //parx*perpx + pary*perpy + parz*perpz == (x1^2 + y1^2 + z1^2) - (x2^2 + y2^2 + z2^2) == d1^2 - d2^2
+            const DOUBLE s_dot_l = dpos*dpos - (*locald1) * (*locald1);
+            localx1++;localy1++;localz1++;locald1++;
+
+            for(int w = 0; w < pair.num_weights; w++){
+                pair.weights1[w].d = *local_w1.weights[w]++;
+            }
+
+            const DOUBLE sqr_s = perpx*perpx + perpy*perpy + perpz*perpz;
+            if(sqr_s >= sqr_smax || sqr_s < sqr_smin) continue;
+
+            const DOUBLE sqr_l = (parx*parx + pary*pary + parz*parz);
+            const DOUBLE sqr_s_dot_l = s_dot_l * s_dot_l;
+            const DOUBLE sqr_mu = sqr_s_dot_l/(sqr_l * sqr_s);
+            const int mubin  = (sqr_mu >= sqr_mumax) ? nmu_bins:(int) (SQRT(sqr_mu)*inv_dmu);
+            DOUBLE s, pairweight;
+            if(need_savg) {
+                s = SQRT(sqr_s);
+            }
+            if(need_weightavg){
+                pair.dx.d = perpx;
```

Comment:
In the future, we might want to think about whether it would be useful to pass (s,mu) to the weighting functions instead of (or in addition to) perp/par[xyz].  I think this would be a fairly lightweight addition; the `pair_struct` is already pretty big, seemingly without much effect on performance, so a few more fields shouldn't matter much.

Reply:
Sure thing - we could provide fields to pass the reduced distance measures as well. Currently, the user would have change these ines of code and pass (s, mu) in any of those 6 (`dx/parx`) fields.